File: RailgunVB/Core/Logging/Log.vb
Imports System.IO
Imports Discord
Imports Discord.WebSocket
Imports RailgunVB.Core.Configuration

Namespace Core.Logging
    
    Public Class Log
        
        Private ReadOnly _config As Configuration.DiscordConfig
        Private WithEvents _client As DiscordShardedClient
        
        Private Const LogDirectory As String = "logs"
        Private ReadOnly _logFilename As String = $"{LogDirectory}/{DateTime.Today.ToString("yyyy-MM-dd")}.log"

        Public Sub New(config As MasterConfig, client As DiscordShardedClient)
            _config = config.DiscordConfig
            _client = client
            
            If Not (Directory.Exists(LogDirectory)) Then Directory.CreateDirectory(LogDirectory)
        End Sub
        
        Public Async Function LogToBotLogAsync(entry As String, type As BotLogType, 
                                               Optional pingMaster As Boolean = False, 
                                               Optional filename As String = Nothing) As Task
            If _config.MasterGuildId = 0 Then Return
            
            Dim guild As IGuild = Await CType(_client, IDiscordClient).GetGuildAsync(_config.MasterGuildId)
            
            Select type
                Case BotLogType.Common
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.Common, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.AudioChord
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.AudioChord, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.CommandManager
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.CommandMngr, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.GuildManager
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.GuildMngr, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.MusicManager
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.MusicMngr, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.MusicPlayer
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.MusicPlayer, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.TaskScheduler
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.TaskSch, entry, pingMaster, filename)
                    Exit Select
                Case BotLogType.TimerManager
                    Await SendBotLogAsync(guild, type, _config.BotLogChannels.TimerMngr, entry, pingMaster, filename)
                    Exit Select
            End Select
        End Function
        
        Public Async Function LogToConsoleAsync(message As LogMessage) As Task Handles _client.Log
            Select message.Severity
                Case LogSeverity.Info
                    Console.ForegroundColor = ConsoleColor.Blue
                    Exit Select
                Case LogSeverity.Critical
                    Console.ForegroundColor = ConsoleColor.DarkRed
                    Exit Select
                Case LogSeverity.Error
                    Console.ForegroundColor = ConsoleColor.Red
                    Exit Select
                Case LogSeverity.Warning
                    Console.ForegroundColor = ConsoleColor.Yellow
                    Exit Select
                Case Else
                    Console.ForegroundColor = ConsoleColor.Magenta
                    Exit Select
            End Select
            
            Console.WriteLine(message.ToString())
            Await WriteToLogFileAsync(message.ToString())
        End Function
        
        Private Async Function SendBotLogAsync(guild As IGuild, type As BotLogType, channelId As ULong,
                                               entry As String, pingMaster As Boolean, filename As String) As Task
            Dim output As String = 
                $"[ {DateTime.Now.ToString("HH:mm:ss")} ] || { _ 
                    If(Not (type = BotLogType.Common), $"{type} ||", "")} {entry}"
            
            Try
                Dim tc As ITextChannel
                
                If Not (channelId = 0) Then
                    tc = Await guild.GetTextChannelAsync(channelId)
                ElseIf Not (_config.BotLogChannels.Common = 0) Then
                    tc = Await guild.GetTextChannelAsync(_config.BotLogChannels.Common)
                Else
                    Return
                End If
                
                Dim pingMasterStr As String = If(pingMaster, $"<@!{_config.MasterAdminId}>", String.Empty)
                
                If Not (String.IsNullOrEmpty(filename))
                    Await tc.SendFileAsync(filename, $"Error! Refer to file, {pingMasterStr}!")
                    File.Delete(filename)
                    Return
                End If
                
                Await tc.SendMessageAsync(Format.Code(output) + pingMasterStr)
            Catch
            End Try
            
            Await WriteToLogFileAsync(output)
        End Function
        
        Private Async Function WriteToLogFileAsync(entry As String) As Task
            Using logFile = File.AppendText(_logFilename)
                Await logFile.WriteLineAsync(entry)
            End Using
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Logging/Log.cs
using System;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Discord;
using Discord.WebSocket;
using Railgun.Core.Configuration;
using Railgun.Core.Utilities;

namespace Railgun.Core.Logging
{
    public class Log
    {
        private const string LogDirectory = "logs";

        private readonly MasterConfig _config;
        private readonly DiscordShardedClient _client;

        public Log(MasterConfig config, DiscordShardedClient client) {
            _config = config;
            _client = client;

            if (!Directory.Exists(LogDirectory)) Directory.CreateDirectory(LogDirectory);

            _client.Log += LogToConsoleAsync;
        }

        private async Task WriteToLogFileAsync(string entry) {
            using (var logfile = File.AppendText($"{LogDirectory}/{DateTime.Today.ToString("yyyy-MM-dd")}.log")) {
                await logfile.WriteLineAsync(entry);
            }
        }

        private async Task SendBotLogAsync(IGuild guild, BotLogType logType, ulong logChannelId, string entry, bool pingMaster) {
            var output = new StringBuilder()
                .AppendFormat("[ {0} ] || {1} {2}", 
                    DateTime.Now.ToString("HH:mm:ss"),
                    logType != BotLogType.Common ? $"{logType} ||" : "",
                    entry);

            try {
                ITextChannel tc;

                if (logChannelId != 0) tc = await guild.GetTextChannelAsync(logChannelId);
                else if (_config.DiscordConfig.BotLogChannels.Common != 0)
                    tc = await guild.GetTextChannelAsync(_config.DiscordConfig.BotLogChannels.Common);
                else return;

                var pingMasterStr = pingMaster ? $"<@!{_config.DiscordConfig.MasterAdminId}>": string.Empty;

                if (output.Length > 1950) {
                    await CommandUtils.SendStringAsFileAsync(tc, "UnobservedTask.log", $"Error! Refer to file {pingMasterStr}!", includeGuildName:false);
                    
                    return;
                }

                await tc.SendMessageAsync(Format.Code(output.ToString()) + pingMasterStr);
            } catch { }

            await WriteToLogFileAsync(output.ToString());
        }

        public async Task LogToBotLogAsync(string entry, BotLogType logType, bool pingMaster = false) {
            if (_config.DiscordConfig.MasterGuildId == 0) return;

            var guild = await ((IDiscordClient)_client).GetGuildAsync(_config.DiscordConfig.MasterGuildId);

            switch (logType) {
                case BotLogType.Common:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.Common, entry, pingMaster);
                    break;
                case BotLogType.AudioChord:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.AudioChord, entry, pingMaster);
                    break;
                case BotLogType.CommandManager:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.CommandMngr, entry, pingMaster);
                    break;
                case BotLogType.GuildManager:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.GuildMngr, entry, pingMaster);
                    break;
                case BotLogType.MusicManager:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.MusicMngr, entry, pingMaster);
                    break;
                case BotLogType.MusicPlayer:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.MusicPlayer, entry, pingMaster);
                    break;
                case BotLogType.TaskScheduler:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.TaskSch, entry, pingMaster);
                    break;
                case BotLogType.TimerManager:
                    await SendBotLogAsync(guild, logType, _config.DiscordConfig.BotLogChannels.TimerMngr, entry, pingMaster);
                    break;
            }
        }

        public async Task LogToConsoleAsync(LogMessage message) {
            switch (message.Severity) {
                case LogSeverity.Info:
                    Console.ForegroundColor = ConsoleColor.Blue;
                    break;
                case LogSeverity.Critical:
                    Console.ForegroundColor = ConsoleColor.DarkRed;
                    break;
                case LogSeverity.Error:
                    Console.ForegroundColor = ConsoleColor.Red;
                    break;
                case LogSeverity.Warning:
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    break;
                default:
                    Console.ForegroundColor = ConsoleColor.Magenta;
                    break;
            }

            Console.WriteLine(message.ToString());

            await WriteToLogFileAsync(message.ToString());
        }
    }
}